src/table_builder.cc

```diff
@@ -115,16 +106,24 @@ void TitanTableBuilder::Add(const Slice& key, const Slice& value) {
                         index.file_number, get_status.ToString().c_str());
       }
     }
-    if (builder_unbuffered()) {
-      base_builder_->Add(key, value);
-    } else {
-      std::unique_ptr<BlobFileBuilder::BlobRecordContext> ctx =
-          NewCachedRecordContext(ikey, value);
-      blob_builder_->AddSmall(std::move(ctx));
-    }
+    AddSmallToTableAdaptively(ikey, value);
   } else {
-    assert(builder_unbuffered());
-    base_builder_->Add(key, value);
+    // Mainly processing kTypeMerge and kTypeBlobIndex in both flushing and compaction.
+    AddSmallToTableAdaptively(ikey, value);
+  }
+}
+
+void TitanTableBuilder::AddSmallToTableAdaptively(const ParsedInternalKey& ikey, const Slice& value){
+  if (builder_unbuffered()) {
+    // We can directly append this into SST safely, without disorder issue.
+    // Only when base_builder_ is in unbuffered state
+    base_builder_->Add(ikey.user_key, value);
+  } else {
+    // We have to let builder to cache this KV pair, and it will be returned
+    // when state changed
```

Comment:
```suggestion
    // We have to let builder to cache this KV pair, and it will be flushed to base table
    // when the state changes to unbuffered
```